### PR TITLE
Improve how we handle our binaries

### PR DESCRIPTION
**Pull request by tduffield**
* All binaries are now self-documenting!
  * All binaries have a '--help' function
  * The contents of the '--help' function are now automatically
    reflected in the README
* Added bats testing framework for binaries.

Signed-off-by: Tom Duffield <tom@chef.io>

### CODE REVIEW

File: bin/citadel [hunk 1]
```diff
@@ -16,37 +16,53 @@
 # limitations under the License.
 #
 
+export CITADEL_PROFILE="${CITADEL_PROFILE:-$AWS_PROFILE}"
+export CITADEL_BUCKET="${CITADEL_BUCKET:-${AWS_PROFILE}-citadel}"
+
+function enforce_requirements() {
+  # Make sure the AWS CLI is installed
+  which aws &>/dev/null
```
[the remark this answers is not in the document]
Comment (tduffield) [reply bin/citadel]: That's not quite right:

```
▶ which aws >/dev/null &2>1
[1] 24112
[1]  + 24112 done       which aws > /dev/null
```

[the remark this answers is not in the document]
Comment (tduffield) [reply bin/citadel]: vs this: 

```
github.com/chef/ci-studio-common  tduffield/self-documenting-bins ✗                                                                                                       7d ◒  ⍉
▶ which aws &>/dev/null

github.com/chef/ci-studio-common  tduffield/self-documenting-bins ✗                                                                                                        7d ◒
▶ echo $?
0

github.com/chef/ci-studio-common  tduffield/self-documenting-bins ✗                                                                                                        7d ◒
▶ which dontexist &>/dev/null

github.com/chef/ci-studio-common  tduffield/self-documenting-bins ✗                                                                                                       7d ◒  ⍉
▶ echo $?
1
```

File: bin/citadel [hunk 2]
```diff
@@ -16,37 +16,53 @@
...
+  if [[ $? -ne 0 ]]; then
+    echo "${0##*/}: Unable to find 'aws' executable."
+    echo "Make sure you have the aws cli installed. Try 'install-tool aws'."
+    exit 2
+  fi
+
+  # Make sure the necessary profile is configured
+  aws-configure is-configured $CITADEL_PROFILE &>/dev/null
```
[the remark this answers is not in the document]
Comment (tduffield) [reply bin/citadel]: not quite. see notes above.

File: bin/citadel (hunk 1 above)
[the remark this answers is not in the document]
Comment (afiune) [reply bin/citadel]: tenor-230909501

File: bin/hab-studio [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    * Add ~/.netrc entry for ACCOUNT in Studio.
+    * Make GITHUB_TOKEN environment variable available in Studio.
+DOC
+}
+
+function configure_github_account() {
+  if [[ -z "$1" || "$1" == "-h" || "$1" == "--help" ]];
+  then
+    configure_github_account_usage
+    exit 0
```
Comment (afiune) [inline bin/hab-studio]: I think this exit code is fine for cases `-h` and `--help` but not when the user doesn't provide a sub-command, in that case an `exit 1` might be better.

File: bin/hab-studio [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  then
+    echo "${0##*/}: Unable to find the required environment variable: $account_token."
+    echo "Make sure it is specified."
+    exit 2
+  fi
+
+  # Add credentials to ~/.netrc
+  if ! grep -w "machine github\.com" $HOME/.netrc > /dev/null
+  then
+    echo -e "machine github.com\n  login ${!account_token}" > $HOME/.netrc
```
Comment (afiune) [inline bin/hab-studio]: I wonder if we should override this file. Just a thought. 🤔

File: bin/hab-studio [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+case "$1" in
+  "configure-github-account")
+    configure_github_account $2
+    ;;
+  "cleanup")
+    cleanup
+    ;;
+  "-h"|"--help"|"")
```
Comment (afiune) [inline bin/hab-studio]: It is actually now a big deal, I guess, but I've seen many commands that if you don't provide an option, they all return an error code. But if you provide the flags to display the help, then you will get no error code. So maybe separate that case from here?

File: bin/hab-studio (hunk 3 above)
Comment (tduffield) [reply bin/hab-studio]: fair point :)

File: bin/hab-studio (hunk 4 above)
Comment (tduffield) [reply bin/hab-studio]: I'll noodle on that.

File: bin/hab-studio (hunk 5 above)
Comment (tduffield) [reply bin/hab-studio]: sure. I think that makes sense.

File: bin/hab-studio (hunk 4 above)
Comment (tduffield) [reply bin/hab-studio]: I think the answer here is "yes" - we want to control this file, especially since it's in the studio.

**Comment by tduffield** [comment]
@afiune when I merge this, all the project impact by #56 will break. We should make sure that the effect projects are updated.

**Pull request merged**